File: native/utils/playMediaId.ts
/**
 * The ONE parser for the "play:" media-id grammar shared with the Kotlin patch
 * (Android Auto). A queue item's id is:
 *
 *     play:<itemId>[::<episodeId>][@@<seconds>]
 *
 * Two JS consumers exist and they see the id in slightly different shapes, so
 * this parser handles both via `opts.hasPrefix`:
 *
 *   - reconcileWithNativePlayer (usePlaybackStore) reads the mediaId straight
 *     off the native queue item, so the "play:" prefix is still attached
 *     (hasPrefix: true). It uses only itemId/episodeId — native already owns the
 *     live position, so the @@seconds suffix is parsed but ignored by that
 *     caller (adoption never seeks).
 *
 *   - the RemotePlayId handler (playbackService) receives the id with ONLY the
 *     "play:" prefix stripped by the native side (`mediaId.removePrefix("play:")`
 *     in onAddMediaItems / the cold-start handoff) — the "::episodeId" and any
 *     "@@<seconds>" stay INLINE in the single forwarded "id" string, i.e.
 *     grammar "<itemId>[::episodeId][@@seconds]" (hasPrefix: false). This parser
 *     does the "@@" and "::" splits, and the caller seeks to bookmarkSeconds
 *     when present and > 0.
 *
 * The grammar contract is pinned in
 * __tests__/contracts/nativeBridgeShapes.test.ts (two-sided with the patch) and
 * unit-tested in __tests__/utils/playMediaId.test.ts.
 */
export interface ParsedPlayMediaId {
  itemId: string;
  /** undefined when there is no "::episodeId" segment (or it is empty). */
  episodeId?: string;
  /**
   * Raw Number() parse of the "@@<seconds>" suffix:
   *   - undefined when the "@@" suffix is absent,
   *   - NaN when the suffix is non-numeric (e.g. "@@later"),
   *   - 0 for an empty suffix ("@@").
   * It is deliberately NOT normalised here — every seek caller already guards
   * with `!isNaN(bookmarkSeconds) && bookmarkSeconds > 0`, so a malformed suffix
   * simply skips the seek rather than being silently coerced to a valid number.
   */
  bookmarkSeconds?: number;
}

export function parsePlayMediaId(
  id: string,
  opts?: { hasPrefix?: boolean }
): ParsedPlayMediaId {
  let raw = id ?? "";
  if (opts?.hasPrefix && raw.startsWith("play:")) {
    raw = raw.slice("play:".length);
  }
  // FIRST-occurrence splits, mirroring the Kotlin parser exactly (its
  // substringBefore/substringAfter both cut at the FIRST delimiter): everything
  // after the first "@@" is the bookmark suffix, everything after the first "::"
  // is the episode id. String.split() would instead cut at EVERY occurrence and
  // diverge from Kotlin on an id carrying a repeated delimiter
  // (e.g. "a@@1@@2" → suffix "1@@2" here vs "1" under split; "a::e::x" →
  // episode "e::x" here vs "e" under split).
  // Take the "@@<seconds>" bookmark/position suffix off first, so the
  // "::episodeId" split never sees it.
  const atIdx = raw.indexOf("@@");
  const main = atIdx === -1 ? raw : raw.slice(0, atIdx);
  const bookmarkStr = atIdx === -1 ? undefined : raw.slice(atIdx + 2);
  const colonIdx = main.indexOf("::");
  const itemId = colonIdx === -1 ? main : main.slice(0, colonIdx);
  // Empty ("a::") or missing episode segment both collapse to undefined, which
  // is the shape startPlayback expects (no episode).
  const episodeId = colonIdx === -1 ? undefined : main.slice(colonIdx + 2) || undefined;
  const bookmarkSeconds = bookmarkStr !== undefined ? Number(bookmarkStr) : undefined;
  return { itemId, episodeId, bookmarkSeconds };
}

File: native/automotive/src/main/java/com/tomesonic/app/automotive/media/PlayMediaId.kt
package com.tomesonic.app.automotive.media

/**
 * The media-id grammar, frozen across all four clients (ARCHITECTURE.md §4.1):
 *
 * ```
 * play:<itemId>[::<episodeId>][@@<seconds>]
 * ```
 *
 * Examples, and the whole of the grammar:
 *
 * ```
 * play:li_abc                    // a book
 * play:li_abc::ep_1              // a podcast episode
 * play:li_abc@@1234.5            // a book, resumed at an absolute second
 * play:li_abc::ep_1@@0           // an episode, explicitly from the start
 * ```
 *
 * A port of the patched RNTP MusicService's `parsePlayMediaId` (L353–369),
 * mirrored on the JS side by `native/utils/playMediaId.ts`. Both split at the
 * FIRST delimiter, and that is load-bearing: `a@@1@@2` has the suffix `1@@2`
 * (not a bookmark of 1) and `a::e::x` has the episode `e::x` (not `e`). A
 * `split()`-based parser looks equivalent and diverges on exactly those ids.
 *
 * The one documented difference from the JS parser: an EMPTY episode segment
 * (`play:a::`) reads as `""` here and as `undefined` there. Kotlin's is the
 * older shape and stays, because it is what the shipped Auto service produces
 * and consumes; every consumer must therefore treat a blank episode id as
 * absent, which is what [PlayMediaId.episodeOrNull] exists for — an empty
 * episode id sent to ABS would build `/api/items/{id}/play/` and 404.
 */
data class PlayMediaId(
    val itemId: String,
    /** May be `""` when the id carried an empty `::` segment — see [episodeOrNull]. */
    val episodeId: String?,
    /** The absolute-seconds override some ids carry; null when absent OR unparseable. */
    val bookmarkSeconds: Double?
) {

    /** The episode id a request may actually use: blank collapses to null. */
    fun episodeOrNull(): String? = episodeId?.takeIf { it.isNotBlank() }

    /** Round-trips to the id this was parsed from, for every well-formed input. */
    fun format(): String = format(itemId, episodeId, bookmarkSeconds)

    companion object {

        const val PREFIX = "play:"

        /** The two delimiters, named once so a call site can't misspell one. */
        const val EPISODE_SEPARATOR = "::"
        const val BOOKMARK_SEPARATOR = "@@"

        fun isPlayId(mediaId: String): Boolean = mediaId.startsWith(PREFIX)

        /**
         * Parses the grammar above.
         *
         * `removePrefix` is a no-op when `play:` is absent, matching every call
         * site (each guards with [isPlayId] first) and matching the JS parser's
         * `hasPrefix` option: a stripped-form id parses as a bare item id.
         *
         * Never throws and never rejects: a malformed id yields an empty
         * [itemId], which is what the caller rejects. A browse callback that
         * threw here would leave the controller hanging on an unset future.
         */
        fun parse(mediaId: String): PlayMediaId {
            val body = mediaId.removePrefix(PREFIX)
            val raw = body.substringBefore(BOOKMARK_SEPARATOR)
            // "@@" with nothing (or nothing numeric) after it is NOT a bookmark
            // of 0 — `"".toDoubleOrNull()` is null, and the JS side guards the
            // same footgun (`Number("")` is 0 there).
            val bookmarkSeconds = body.substringAfter(BOOKMARK_SEPARATOR, "").toDoubleOrNull()
            val itemId = raw.substringBefore(EPISODE_SEPARATOR)
            val episodeId =
                if (raw.contains(EPISODE_SEPARATOR)) raw.substringAfter(EPISODE_SEPARATOR) else null
            return PlayMediaId(itemId, episodeId, bookmarkSeconds)
        }

        /**
         * Builds an id. The inverse of [parse] for every id this module emits.
         *
         * A blank [episodeId] is omitted rather than written as an empty `::`
         * segment: the browse tree only ever has one or the other, and emitting
         * the ambiguous form would push the JS/Kotlin divergence above onto the
         * wire. [seconds] is interpolated exactly as the donor's handoff builds
         * it (`"$raw@@$absSec"`), so a Double's own `toString` is the format.
         */
        fun format(itemId: String, episodeId: String? = null, seconds: Double? = null): String {
            val episode = episodeId?.takeIf { it.isNotBlank() }
            val sb = StringBuilder(PREFIX).append(itemId)
            if (episode != null) sb.append(EPISODE_SEPARATOR).append(episode)
            if (seconds != null) sb.append(BOOKMARK_SEPARATOR).append(seconds)
            return sb.toString()
        }
    }
}
